File: 0229-majority-element-ii/0229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        
        cnt1=0
        cnt2=0
        ele1=None
        ele2=None

        for i in range(len(nums)):
            if cnt1==0 and ele2!=nums[i]:
                ele1=nums[i]
                cnt1=1
            elif cnt2==0 and ele1!=nums[i]:
                ele2=nums[i]
                cnt2=1
            elif ele1==nums[i]:
                cnt1+=1
            elif ele2==nums[i]:
                cnt2+=1
            else:
                cnt1-=1
                cnt2-=1
            
        array=[]
        cnt1=0
        cnt2=0
        for i in range(len(nums)):
            if ele1==nums[i]:
                cnt1+=1
            if ele2==nums[i]:
                cnt2+=1

        if cnt1>(len(nums)//3):
            array.append(ele1)
        if cnt2>(len(nums)//3):
            array.append(ele2)

        return array


        

File: 0229-majority-element-ii/0229-majority-element-ii.java
class Solution {
    public List<Integer> majorityElement(int[] nums) {
        int cnt1=0;
        int cnt2=0;
        int ele1=0;
        int ele2=0;
        for(int num:nums){
            if(cnt1==0 && ele2!=num){
                cnt1++;
                ele1=num;
            }else if(cnt2==0 && ele1!=num){
                cnt2++;
                ele2=num;
            }else if(ele1==num){
                cnt1++;
            }else if(ele2==num){
                cnt2++;
            }else {
                cnt1--;
                cnt2--;
            }
        }
            List<Integer> array=new ArrayList<>();
            cnt1=0;
            cnt2=0;
            int n=nums.length;
            for(int num:nums){
                if(ele1==num){
                    cnt1++;
                }else if(ele2==num){
                    cnt2++;
                }
            }
            if(cnt1>n/3){
                array.add(ele1);
            }
            if(cnt2>n/3){
                array.add(ele2);
            }
            return array;
    }
}